### ISSUE TITLE
The correct loss combination for bm training

### ISSUE THREAD

**Issue by sdsajk** [cvlab-stonybrook/DewarpNet#31]
Hi! I have been using this repo for months and have trained many models. For bm training,  I have trained many models but none of them are usable. I noticed that for loss of bm model, there are comment session of SSIM loss. Should we add this to bm training? What is the correct setup for the loss of bm training? I tried 10* l1loss + 0.5*rloss but the performances are terrible.

**Comment by sagniklp** [comment]
Definitely, there's something wrong.
Before I can help you I need to see the results, what do you mean by terrible?
The suggested training procedure is the following:
1. Input is the normalized world coordinate (wc) maps
2. Output is the backward maps normalized to [0,1]
3. Augmentation: tight crop the wc maps with random padding (tight_crop function does that)
4. At first train with only the l1loss then add the rloss.
5. Train with a small set of images first (~10k), after converging on them progressively increase the number of images.

**Comment by sdsajk** [comment]
Thank you so much for quick response! 

What I do so far is basically to reproduce the performance of the trained model you gave(dnetccnl_doc3d.pkl), and I am using the official doc3d dataset.
So far I didn't modify anything yet. I just train with default settings you gave in README.md, and test the result with infer.py( combined with unet_doc3d.pkl). Below are those result. 
![1](https://user-images.githubusercontent.com/54885810/110624402-1f6c2700-81d9-11eb-9793-012ccf2d770d.png)
I think the major problem is that I use "loss = 10* l1loss + 0.5*rloss" directly from beginning (although the loss did converge).

Can you be more specific about the training setting to reproduce the standard model on doc3d dataset? Like how many epochs are needed to finish the training with l1loss.

Sorry for taking your time, but I really want to figure this out.

**Comment by Wanwan1996** [comment]
Where is the augementation operation used?
We used augementation when we separately trained the wc
Augementation was also used in the wc of the joint training phase.
is this correct?
Do I still need to use augementation in the bm stage, as you said above, during joint training, the augementation should be used for cropping before the image is input to bm?
please help me！！！

**Comment by sdsajk** [comment]
@sagniklp I don't notice the mismatch in the code, but when use provided models to infer.py, I can get correct output, so I assume it is not a big problem?
I have training log here if it is helpful.
[dnetccnl_htan_swat3dmini1kbm_l1_noaug_scratch.txt](https://github.com/cvlab-stonybrook/DewarpNet/files/6116139/dnetccnl_htan_swat3dmini1kbm_l1_noaug_scratch.txt)
.

**Comment by sagniklp** [comment]
@sdsajk It's not the mismatch in code recheck the values in the infer.py there may be something wrong in the input.
Training log looks fine to me.

**Comment by sagniklp** [comment]
> Where is the augementation operation used?
> We used augementation when we separately trained the wc
> Augementation was also used in the wc of the joint training phase.
> is this correct?
> Do I still need to use augementation in the bm stage, as you said above, during joint training, the augementation should be used for cropping before the image is input to bm?
> please help me！！！

tight_crop function implements the crop augmentation for bm. What's the issue? 
Please comment in the respective issue topic.

**Comment by sdsajk** [comment]
@sagniklp The problem is that when I use provided model(dnetccnl_doc3d.pkl) with exactly same settings and same inputs in infer.py I can get correct results, so I don't think it's the problem of the input. 

Here are results when I ran infer.py with provided model(dnetccnl_doc3d.pkl), it is completely different with the result of the model I have trained.
![1](https://user-images.githubusercontent.com/54885810/110641036-792a1c80-81ec-11eb-8a19-e208dc7620b6.png)
Provided model was trained from this repo right?

**Comment by sdsajk** [comment]
@sagniklp New update, I know the meaning of "mismatch of values in inputs". I did the separate test on bm itself and it works out. Thank you for pointing that out!

But how do we deal with that mismatch during inference? I tried just normalize the output of wc model but it didn't work.

**Comment by sagniklp** [comment]
check what you're training with and just adjust infer.py accordingly. 
It may happen due to different normalization used in training and testing.
